File: src/main/java/frc/team687/utilities/statespace/StateSpaceObserver.java
package frc.team687.utilities.statespace;

import Jama.Matrix;

public class StateSpaceObserver {

    private StateSpaceGains[] m_gains;
    private int m_selectedGainsIndex;

    private Matrix m_xHat, m_currentInput;

    public StateSpaceObserver(StateSpaceGains[] gains, Matrix initialState) {
        this.m_gains = gains;
        this.m_selectedGainsIndex = 0;

        this.m_xHat = initialState;
    }

    public void setGainsIndex(int index) {
        if (index > this.m_gains.length - 1) {
            this.m_selectedGainsIndex = index;
        } else {
            this.m_selectedGainsIndex = this.m_gains.length - 1;
        }
    }

    public Matrix newStateEstimate(Matrix u, Matrix y) {
        StateSpaceGains gains = this.m_gains[m_selectedGainsIndex];

        /*
        x_hat[k+1] = Ax_hat[k] + Bu[k] + L(y[k] - y_hat[k])
        x_hat[k+1] = Ax_hat[k] + Bu[k] + L(y[k] - C(x_hat[k]))
        x_hat[k+1] = Ax_hat[k] + Bu[k] + Ly[k] - LCx_hat[k]
        x_hat[k+1] = (A - LC)x_hat[k] + Bu[k] + Ly[k]
        */

        // (A - LC)x_hat[k]
        Matrix xHatTerm = gains.A.minus(gains.L.times(gains.C)).times(this.m_xHat);
        // Bu[k]
        Matrix Bu = gains.B.times(u);
        // Ly[k]
        Matrix Ly = gains.L.times(y);

        this.m_xHat = xHatTerm.plus(Bu.plus(Ly));
        return this.m_xHat;
    }

    public Matrix getCurrentStateEstimate() {
        return this.m_xHat;
    }

    public void setXHat(Matrix xHat) {
        this.m_xHat = xHat;
    }

}

File: src/main/python/utilities/state_space/state_space_observer.py
from utilities.state_space.state_space_gains import GainsList


class StateSpaceObserver(object):

    def __init__(self, gains: GainsList, x_hat_initial):
        self.gains = gains
        self.gains_index = 0
        self.current_gains = self.gains.get_gains(self.gains_index)

        self.x_hat = x_hat_initial

    def set_index(self, index: int):
        self.gains_index = index
        self.current_gains = self.gains.get_gains(index)

    def update(self, u, y):
        gains = self.current_gains

        self.x_hat = (gains.A - (gains.L * gains.C)) * self.x_hat + gains.B * u + gains.L * y

        return self.x_hat
